os/kernel/sched/sched_resumescheduler.c

```diff
@@ -74,46 +74,6 @@ void sched_resume_scheduler(FAR struct tcb_s *tcb)
 #ifdef CONFIG_SCHED_INSTRUMENTATION
   sched_note_resume(tcb);
 #endif
-
-#ifdef CONFIG_SMP
-  /* NOTE: The following logic for adjusting global IRQ controls were
-   * derived from sched_add_readytorun() and sched_removedreadytorun()
-   * Here, we only handles clearing logic to defer unlocking IRQ lock
-   * followed by context switching.
-   */
-
-  int me = this_cpu();
-
-  /* Adjust global IRQ controls.  If irqcount is greater than zero,
-   * then this task/this CPU holds the IRQ lock
-   */
-
-  if (tcb->irqcount > 0)
-    {
-      /* Do nothing here
-       * NOTE: spin_setbit() is done in sched_add_readytorun()
-       * and sched_remove_readytorun()
-       */
-    }
-
-  /* No.. This CPU will be relinquishing the lock.  But this works
-   * differently if we are performing a context switch from an
-   * interrupt handler and the interrupt handler has established
-   * a critical section.  We can detect this case when
-   * g_cpu_nestcount[me] > 0.
-   */
-
-  else if (g_cpu_nestcount[me] <= 0)
-    {
-      /* Release our hold on the IRQ lock. */
-
-      if ((g_cpu_irqset & (1 << me)) != 0)
-        {
-          spin_clrbit(&g_cpu_irqset, me, &g_cpu_irqsetlock,
-                      &g_cpu_irqlock);
-        }
-    }
-#endif /* CONFIG_SMP */
```

Comment:
You are moving this code piece out to form a new API --> restore_critical_section(). But here you are not invoking that API.
As I checked, there are quite some other places invoking sched_resume_scheduler(). After removing this code piece, the remaining operation left in this API is wrapped by macros, which mean we might be doing nothing here if all macros are disabled.
Could you please check whether a restore_critical_section() is needed here?

Reply:
I have few observations on this. 
First of all, I feel sched_resume_scheduler() is supposed to do only scheduler related work and the irqcount related code was placed wrongly in it. When we took the latest code from nuttx, it seems they have corrected this. 
Secondly, I see that resume_scheduler() API is called in following places. 

os/arch/arm/src/armv7-a/arm_releasepending.c
os/arch/arm/src/armv7-a/arm_cpustart.c
os/arch/arm/src/armv7-a/arm_exit.c
os/arch/arm/src/armv7-a/arm_unblocktask.c 	 	
os/arch/arm/src/armv7-a/arm_reprioritizertr.c
os/arch/arm/src/armv7-a/arm_cpupause.c  
os/arch/arm/src/armv7-a/arm_blocktask.c 

In each of the above cases, the API is called from IRQ context. But now we have moved the irqcount logic to doirq() and so, we would be still performing this check when we are returning from IRQ context. So, I think this is fine.

The only exception is arm_exit(). However, since this code is working well in nuttx and also all our tests seem to be fine, I think we can keep this change for now. If we face any issues in future, we can recheck this part.

Reply:
Thank you for detailed explanation, it make sense to me.
But, could you please add a remark or comment in the code piece here, since it might be a potential checkpoint?

Reply:
I have added a note in the commit msg.